### ISSUE TITLE
Build tool performance improvements

### ISSUE THREAD

**Issue by dgreensp** [meteor/meteor#6058]
**See also this doc:** https://github.com/meteor/meteor/blob/release-1.3/tools/PERFORMANCE.md

This issue tracks build tool performance improvements being worked on in the context of Meteor 1.3.

When talking about build tool performance, the following factors should be taken into account:
- Initial build vs. rebuild: As you'd expect, rebuild takes much less time than initial build, because of caching.  A performance improvement that cuts down initial build time may not affect rebuild time.  Also, there's a distinction between a client-only and a client/server rebuild.  In addition, the _first_ rebuild after starting an app is currently slower than subsequent rebuilds because of how the constraint solver does caching.
- Running from release vs. checkout: Most people run Meteor "from a release," meaning they run the installed `meteor` on their machine (with or without `--release`).  When developing and testing Meteor, we often from "from checkout," meaning we run `./meteor` from the checked-out repo.  These two modes have very different performance characteristics.  When you run from checkout, the core packages are loaded from the repo.  This means that, for example, a lot more files are watched, but the constraint solver doesn't have to do nearly as much work because the core package versions are pinned.  While it's easier to debug and inspect Meteor when running from checkout, when working on performance improvements we should focus on the experience of running from release.
- Size and complexity of app:  One app may be slower to build than another because it is larger or more complex on in any number of dimensions, such as the amount of code to compile or CSS to minify; the compiler plugins used; the size of the dependency graph; or the size in bytes of the packages used.  We should make sure "skeleton" (the empty app) is fast, and also that typical "real apps" are fast.
- Hardware and OS:  Numbers will vary based on your CPU speed and disk speed (especially whether you have an SSD or a spinning disk).  Also, the Windows file system has different characteristics from a Mac or Linux file system.

Even though absolute time measurements will vary from computer to computer and app to app, it's useful to talk in rough numbers about where time is going.
#### Overview of Rebuilding

When a "real app" rebuilds running from release, it may take as much as 10-20 seconds on `1.3-modules-beta.4` (the latest release as of the beginning of this work).  Most of this time goes to "bundling."  The bundler is instrumented using Meteor's profiling library, so if you run `METEOR_PROFILE=1 meteor` in a shell, you'll get a decent overview of where the time is going (under the heading "Rebuild App").

Besides bundling, there are a couple other "hidden" sources of rebuild time which you won't see in the profile, but they are small in comparison:
- App start-up (~1 second): Meteor runs two Node processes, one for the CLI tool and one for the app.  On rebuild, the app is killed and restarted, and the message `=> Meteor server restarted` is displayed once the new app process is live and serving.
- File watcher set-up (hundreds of ms): On rebuild, the tool tears down the old file watchers and sets up new ones.  This set-up is performed in parallel with app start-up, so it is not really noticeable when running from release.  (When running from checkout, there are so many files to watch that it may take several seconds and becomes noticeable.)
#### Potential improvements

These are opportunities for half-second to multi-second improvements:

Rebuild:
- [X] Fix linker's broken RAM cache
- [X] Cache code analysis to detect `import` in modules
- [X] Don't stat when writing files
- [X] Upgrade `source-map` NPM package
- [x] Speed up read/stat of NPM dependencies
- [ ] Speed up CSS analysis and concatenation - see https://github.com/meteor/meteor/pull/5875
- [x] Change line-number comments in linked files to not inflate file sizes so much (2-3 times?)
- [x] Overwrite existing bundle on build (for far less file writing)
- [ ] Further investigate source map performance

Initial build:
- [ ] "Updating package catalog…" seems to take a long time; why? (~5 seconds?)
- [x] Implement constraint solver "fast path" if `.meteor/versions` is valid (5-8 seconds)
- [x] Optimize package version parsing/sorting when loading catalog data (1-2 seconds out of the above?)
- [ ] To reproduce: Rebuild gradually gets slower over time and becomes more like 30+ seconds?

**Comment by markoshust** [comment]
Please see my notes at https://forums.meteor.com/t/help-us-test-build-times-in-meteor-1-3/15031/26?u=markoshust -- thought it may be helpful.

Cheers,
Mark

**Comment by dgreensp** [comment]
@markoshust, thanks!  It would be amazing if you could post a repository of the JSX files that take a full minute to rebuild!

**Comment by markoshust** [comment]
Ok -- I put together a skeleton layout and I had... 3 second reload times! Give me a few, I need to find out where this delay is originating from. I'm doing some additional tests to try to find the bottleneck.

**Comment by markoshust** [comment]
Ok, I was able to replicate the origination of the slow build times.

About 10 or so seconds was from having the `jsx` meteor package installed -- this is no longer needed. However, the much bigger issue is caused by how imports are handled.

Here is a sample repo to replicate the slow build times.

Fast (~30 seconds initial build, <=7 sec reload times)
https://github.com/markoshust/meter13-matui/tree/fast

Slow (~55 seconds initial build, ~30 sec reload times)
https://github.com/markoshust/meter13-matui/tree/slow

This commit is the only change between the two branches. The one on the left is the fast build, the one on the right is the slow build:
https://github.com/markoshust/meter13-matui/commit/9c9f3535f74e29b518e800ed76b40a1d3c7c74d9

I could switch to the left format, as it looks like the docs at www.material-ui.com use the longer format (more explicit is better?). However, I'm not sure if the right format is still more commonplace.

Mark

**Comment by dgreensp** [comment]
Hi Mark,

Thanks for that.

You definitely want the code on the left.  It's not just syntax; there's a big difference in what code is included in your app and shipped to the client!

The new module-based build system has the benefit that modules you never depend on are not included in your app -- it's like dead code elimination at the file (module) level.  By depending on the module `material-ui/lib`, which is a [module](https://github.com/callemall/material-ui/blob/e12a66c2c2db5f3109111d2273cfd2dc54d02930/src/index.js) that depends on a few dozen other modules, your app takes a dependency on all of them.

I'll use the 30-second reload times as one of my benchmarks; that's pretty slow.

**Comment by dgreensp** [comment]
@markoshust, are you able to run a branch of Meteor from the repository?  If so, I would ask you to test the 30-second reload case on a branch I'm working on.

**Comment by markoshust** [comment]
@dgreensp thanks -- I figured so, so I actually went ahead and rewrote everything in the left format. Thank for your the detailed explanation -- glad that is all cleared up :)

Sure, just let me know how to run off a branch and I'll be very glad to test it out. Here are my final figures with my entire app being rewritten for 1.3:

Initial load:
```
real    0m42.627s
user    0m37.383s
sys     0m3.876s
```

Reloads now take roughly 25 seconds -- better than the previous 1min+ figures I was receiving, but still very slow. The same app roughly took 17 seconds to reload in 1.2. I would expect moving a few packages out of the meteor packaging system and into npm, along with using imports exclusively, would be getting way better figures than 17 seconds.

**Comment by dgreensp** [comment]
@markoshust, I'm not sure whether moving code into NPM ought to be a net positive or not.  At the moment, spidering and relinking NPM modules seems to take a while, and it may be less incremental than keeping the code in Meteor packages, but I'm not an expert on the code.

**Comment by markoshust** [comment]
@dgreensp here's an update for 1.3-b6

my app's initial load:

```
real	0m36.491s
user	0m32.253s
sys	0m4.035s
```

reloads take 10-12 seconds. wayyyy better (better than times i was getting on 1.2), still probably some ways to go though :+1:

**Comment by markoshust** [comment]
update for app 1.3-b7

initial load:
```
real	0m30.401s
user	0m26.172s
sys	0m3.106s
```

reloads still about 12 seconds.

**Comment by dgreensp** [comment]
Hi @markoshust, can you post the complete output of `METEOR_PROFILE=1 meteor ...`, including two reloads, as a gist?  Thanks!

**Comment by markoshust** [comment]
Sure. Here you go. Before I was running with `time` prefixing my meteor run command. I should have been doing it this way instead :/

```
=> Started proxy.                             
| (#1) Profiling: ProjectContext prepareProjectForBuild
=> Started MongoDB.                           
|  Preparing project                         |
| files.readFile                                                  938 ms (3)
| bundler.readJsImage..............................................11 ms (1)
|    files.readFile                                                 9 ms (11)
|    other bundler.readJsImage                                      3 ms
| JsImage#load....................................................259 ms (1)
|    runJavaScript packages/meteor-env-prod.js                      1 ms (1)
|    runJavaScript packages/underscore.js                          36 ms (1)
|    runJavaScript packages/meteor.js                              35 ms (1)
|    runJavaScript packages/npm-mongo.js                          187 ms (1)
| files.writeFile                                                  26 ms (1)
| ProjectContext prepareProjectForBuild.........................7,623 ms (1)
|    _initializeCatalog                                             4 ms (1)
|    _resolveConstraints........................................5,961 ms (1)
|       bundler.readJsImage........................................33 ms (1)
|          files.readFile                                          25 ms (21)
|          other bundler.readJsImage                                7 ms
|       JsImage#load..............................................414 ms (1)
|          runJavaScript packages/meteor-env-prod.js                7 ms (1)
|          runJavaScript packages/underscore.js                    60 ms (1)
|          runJavaScript packages/meteor.js                        43 ms (1)
|          runJavaScript packages/base64.js                         5 ms (1)
|          runJavaScript packages/ejson.js                         25 ms (1)
|          runJavaScript packages/check.js                         14 ms (1)
|          runJavaScript packages/package-version-parser.js        53 ms (1)
|          runJavaScript packages/logic-solver.js                 127 ms (1)
|          runJavaScript packages/constraint-solver.js             79 ms (1)
|          other JsImage#load                                       1 ms
|       Select Package Versions.................................5,503 ms (1)
|          new CS.Input                                             7 ms (1)
|          Input#loadFromCatalog (sqlite)                       4,251 ms (1)
|          new CS.Solver...........................................48 ms (1)
|             Solver#analyze.......................................48 ms (1)
|                analyze allowed versions                          17 ms (1)
|                analyze reachability                               6 ms (1)
|                analyze constraints                               23 ms (1)
|                analyze pre-releases                               1 ms (1)
|          Solver#getAnswer.....................................1,196 ms (1)
|             new Logic.Solver (MiniSat start-up)                  16 ms (1)
|             require root dependencies                             1 ms (1)
|             generate package variables                          148 ms (1)
|             generate dependency requirements                     90 ms (1)
|             generate constraints                                163 ms (1)
|             pre-solve                                           432 ms (1)
|             minimize unknown_packages                             3 ms (1)
|             minimize conflicts                                  124 ms (1)
|             minimize unanticipated_prereleases                    1 ms (1)
|             calculate previous_root distance costs                8 ms (1)
|             minimize previous_root_incompat                      84 ms (1)
|             minimize previous_root_major                          5 ms (1)
|             minimize previous_root_minor                          6 ms (1)
|             minimize previous_root_patch                         21 ms (1)
|             calculate previous_indirect distance costs            4 ms (1)
|             minimize previous_indirect_incompat                   6 ms (1)
|             minimize previous_indirect_minor                      6 ms (1)
|             minimize previous_indirect_patch                      1 ms (1)
|             lock down important versions                          5 ms (1)
|             minimize total_packages                              66 ms (1)
|             generate version map                                  1 ms (1)
|             other Solver#getAnswer                                5 ms
|       other _resolveConstraints                                  11 ms
|    _downloadMissingPackages.....................................162 ms (1)
|       Isopack.readMetadataFromDirectory.........................103 ms (109)
|          files.exists                                            25 ms (109)
|          files.readFile                                          68 ms (109)
|          other Isopack.readMetadataFromDirectory                 10 ms
|       other _downloadMissingPackages                             59 ms
|    _buildLocalPackages........................................1,494 ms (1)
|       IsopackCache Load local isopack.........................1,481 ms (109)
|          files.stat                                              16 ms (109)
|          Isopack#initFromPath.................................1,359 ms (109)
|             files.realpath                                       79 ms (109)
|             Isopack.readMetadataFromDirectory....................47 ms (109)
|                files.exists                                      10 ms (109)
|                files.readFile                                    27 ms (109)
|                other Isopack.readMetadataFromDirectory           11 ms
|             files.readFile                                      184 ms (329)
|             files.open                                          105 ms (1103)
|             files.read                                          423 ms (1103)
|             files.close                                          84 ms (1103)
|             bundler.readJsImage.................................224 ms (7)
|                files.readFile                                   190 ms (155)
|                other bundler.readJsImage                         33 ms
|             other Isopack#initFromPath                          213 ms
|          Isopack#getStrongOrderedUsedAndImpliedPackages           9 ms (109)
|          other IsopackCache Load local isopack                   98 ms
|       other _buildLocalPackages                                  13 ms
|    _saveChangedMetadata                                           1 ms (1)
| 
| Top leaves:
| Input#loadFromCatalog (sqlite)...........................4,251 ms
| files.readFile...........................................1,442 ms
| pre-solve..................................................432 ms
| files.read.................................................423 ms
| other Isopack#initFromPath.................................213 ms
| generate constraints.......................................163 ms
| generate package variables.................................148 ms
| runJavaScript packages/logic-solver.js.....................127 ms
| minimize conflicts.........................................124 ms
| files.open.................................................105 ms
| 
| (#1) Total: 8,860 ms (ProjectContext prepareProjectForBuild)
| 
| (#2) Profiling: Build App
|  Building the application                  |
| files.readFile                                                    3 ms (3)
| bundler.readJsImage..............................................57 ms (1)
|    files.readFile                                                47 ms (45)
|    other bundler.readJsImage                                     10 ms
| JsImage#load....................................................485 ms (1)
|    runJavaScript packages/meteor-env-prod.js                      1 ms (1)
|    runJavaScript packages/underscore.js                          53 ms (1)
|    runJavaScript packages/meteor.js                              43 ms (1)
|    runJavaScript packages/base64.js                               5 ms (1)
|    runJavaScript packages/ejson.js                               27 ms (1)
|    runJavaScript packages/check.js                               12 ms (1)
|    runJavaScript packages/babel-compiler.js                      12 ms (1)
|    runJavaScript packages/ecmascript.js                           2 ms (1)
|    runJavaScript packages/modules-runtime.js                     13 ms (1)
|    runJavaScript packages/promise.js                             19 ms (1)
|    runJavaScript packages/ecmascript-runtime.js                 104 ms (1)
|    runJavaScript packages/babel-runtime.js                       47 ms (1)
|    runJavaScript packages/random.js                              23 ms (1)
|    runJavaScript packages/tracker.js                             37 ms (1)
|    runJavaScript packages/retry.js                                2 ms (1)
|    runJavaScript packages/id-map.js                               2 ms (1)
|    runJavaScript packages/ddp-common.js                          11 ms (1)
|    runJavaScript packages/diff-sequence.js                        5 ms (1)
|    runJavaScript packages/mongo-id.js                             3 ms (1)
|    runJavaScript packages/ddp-client.js                          62 ms (1)
|    other JsImage#load                                             1 ms
| files.exists                                                     49 ms (4)
| Build App....................................................20,771 ms (1)
|    initFromAppDir                                                 1 ms (1)
|    compiler.compile(the app)..................................2,598 ms (1)
|       compileUnibuild (the app)...............................2,597 ms (3)
|          Isopack#getUnibuildAtArch                                5 ms (189)
|          Isopack#ensurePluginsInitialized.....................1,762 ms (36)
|             JsImage#load......................................1,753 ms (5)
|                runJavaScript packages/meteor-env-prod.js          4 ms (4)
|                runJavaScript packages/underscore.js             185 ms (4)
|                runJavaScript packages/meteor.js.................766 ms (4)
|                   files.exists                                  608 ms (24)
|                   other runJavaScript packages/meteor.js        158 ms
|                runJavaScript packages/base64.js                  10 ms (2)
|                runJavaScript packages/ejson.js                   40 ms (2)
|                runJavaScript packages/check.js                   24 ms (2)
|                runJavaScript packages/babel-compiler.js..........46 ms (2)
|                   files.exists                                   26 ms (2)
|                   other runJavaScript packages/babel-compiler.js 20 ms
|                runJavaScript packages/compile-ecmascript_plugin.js 1 ms (1)
|                runJavaScript packages/ecmascript.js               1 ms (1)
|                runJavaScript packages/modules-runtime.js         11 ms (1)
|                runJavaScript packages/promise.js.................20 ms (1)
|                   files.exists                                    2 ms (2)
|                   other runJavaScript packages/promise.js        18 ms
|                runJavaScript packages/ecmascript-runtime.js      88 ms (1)
|                runJavaScript packages/babel-runtime.js           33 ms (1)
|                runJavaScript packages/random.js                  21 ms (1)
|                runJavaScript packages/caching-compiler.js........49 ms (1)
|                   files.exists                                    4 ms (8)
|                   other runJavaScript packages/caching-compiler.js 45 ms
|                runJavaScript packages/tracker.js                 23 ms (1)
|                runJavaScript packages/htmljs.js                  20 ms (1)
|                runJavaScript packages/html-tools.js             108 ms (1)
|                runJavaScript packages/blaze-tools.js             17 ms (1)
|                runJavaScript packages/minifiers-js.js           144 ms (2)
|                runJavaScript packages/spacebars-compiler.js      39 ms (1)
|                runJavaScript packages/templating-tools.js        18 ms (1)
|                runJavaScript packages/caching-html-compiler.js   15 ms (1)
|                runJavaScript packages/compileTemplatesBatch_plugin.js 1 ms (1)
|                runJavaScript packages/minifiers-css.js           38 ms (1)
|                runJavaScript packages/minifyStdCSS_plugin.js     24 ms (1)
|                runJavaScript packages/minifyStdJS_plugin.js       2 ms (1)
|                other JsImage#load                                 2 ms
|             other Isopack#ensurePluginsInitialized                9 ms
|          files.readdir                                          113 ms (555)
|          files.stat                                             252 ms (1793)
|          files.realpath                                          64 ms (278)
|          files.readFile                                         122 ms (363)
|          sha1                                                     8 ms (363)
|          other compileUnibuild (the app)                        272 ms
|       other compiler.compile(the app)                             1 ms
|    compiler.lint(the app).........................................8 ms (1)
|       Isopack#getUnibuildAtArch                                   3 ms (126)
|       other compiler.lint(the app)                                5 ms
|    Isopack#getUnibuildAtArch                                      3 ms (189)
|    bundler.bundle..makeClientTarget..........................14,841 ms (2)
|       Target#make............................................14,840 ms (2)
|          Isopack#getUnibuildAtArch                               25 ms (1770)
|          Target#_runCompilerPlugins...........................4,256 ms (2)
|             files.stat                                            2 ms (2)
|             files.realpath                                      149 ms (174)
|             Isopack#ensurePluginsInitialized..................1,683 ms (378)
|                JsImage#load...................................1,679 ms (2)
|                   runJavaScript packages/underscore.js           94 ms (2)
|                   runJavaScript packages/meteor.js...............91 ms (2)
|                      files.exists                                 3 ms (12)
|                      other runJavaScript packages/meteor.js      88 ms
|                   runJavaScript packages/base64.js                9 ms (2)
|                   runJavaScript packages/ejson.js                38 ms (2)
|                   runJavaScript packages/check.js                24 ms (2)
|                   runJavaScript packages/babel-compiler.js.......23 ms (2)
|                      files.exists                                 1 ms (2)
|                      other runJavaScript packages/babel-compiler.js 22 ms
|                   runJavaScript packages/ecmascript.js            3 ms (2)
|                   runJavaScript packages/babel-runtime.js        42 ms (2)
|                   runJavaScript packages/ecmascript-runtime.js  282 ms (2)
|                   runJavaScript packages/promise.js..............44 ms (2)
|                      files.exists                                 3 ms (4)
|                      other runJavaScript packages/promise.js     41 ms
|                   runJavaScript packages/random.js               42 ms (2)
|                   runJavaScript packages/caching-compiler.js.....87 ms (2)
|                      files.exists                                 4 ms (16)
|                      other runJavaScript packages/caching-compiler.js 83 ms
|                   runJavaScript packages/CosmosBrowserify_plugin.js.506 ms (1)
|                      files.exists                                 2 ms (4)
|                      other runJavaScript packages/CosmosBrowserify_plugin.js 505 ms
|                   runJavaScript packages/modules-runtime.js      13 ms (1)
|                   runJavaScript packages/compileCoffeescript_plugin.js.376 ms (1)
|                      files.exists                                 1 ms (3)
|                      other runJavaScript packages/compileCoffeescript_plugin.js 375 ms
|                   other JsImage#load                              2 ms
|                other Isopack#ensurePluginsInitialized             5 ms
|             sha1                                                 13 ms (465)
|             Isopack#getUnibuildAtArch                            37 ms (1714)
|             plugin ecmascript.................................2,170 ms (4)
|                Babel.compile                                    532 ms (310)
|                sha1                                               8 ms (310)
|                other plugin ecmascript                        1,631 ms
|             plugin templating....................................10 ms (2)
|                wrapped.fs.readFileSync                            3 ms (4)
|                other plugin templating                            7 ms
|             plugin meteor                                         6 ms (2)
|             plugin cosmos:browserify..............................7 ms (2)
|                wrapped.fs.readFileSync                            2 ms (1)
|                other plugin cosmos:browserify                     5 ms
|             plugin coffeescript...................................6 ms (2)
|                wrapped.fs.readFileSync                            1 ms (2)
|                other plugin coffeescript                          4 ms
|             other Target#_runCompilerPlugins                    171 ms
|          Target#_emitResources...............................10,146 ms (2)
|             PackageSourceBatch.computeJsOutputFilesMap........9,577 ms (2)
|                ImportScanner#_getInstallPath                    104 ms (1416)
|                ImportScanner#_findImportedModuleIdentifiers   4,005 ms (1344)
|                ImportScanner#_tryToResolveImportedPath........4,669 ms (4700)
|                   files.stat                                  3,275 ms (21654)
|                   ImportScanner#_readFile.......................286 ms (724)
|                      files.readFile                             219 ms (724)
|                      sha1                                        13 ms (712)
|                      other ImportScanner#_readFile               54 ms
|                   ImportScanner#_getInstallPath                   3 ms (42)
|                   other ImportScanner#_tryToResolveImportedPath 1,105 ms
|                ImportScanner#_readFile..........................456 ms (1098)
|                   files.readFile                                352 ms (1098)
|                   sha1                                           23 ms (1098)
|                   other ImportScanner#_readFile                  81 ms
|                other PackageSourceBatch.computeJsOutputFilesMap 342 ms
|             PackageSourceBatch#getResources.....................552 ms (176)
|                PackageSourceBatch#_linkJS.......................546 ms (176)
|                   sha1                                            4 ms (176)
|                   files.readFile                                278 ms (92)
|                   other PackageSourceBatch#_linkJS              264 ms
|                other PackageSourceBatch#getResources              6 ms
|             other Target#_emitResources                          16 ms
|          Target#minifyJs                                         29 ms (2)
|          ClientTarget#minifyCss.................................278 ms (2)
|             mergeCss............................................273 ms (2)
|                CssTools.parseCss                                 92 ms (68)
|                CssTools.mergeCssAsts.............................15 ms (2)
|                   CssTools.rewriteCssUrls                        10 ms (68)
|                   other CssTools.mergeCssAsts                     5 ms
|                CssTools.stringifyCss                            113 ms (2)
|                composing source maps                             39 ms (2)
|                other mergeCss                                    14 ms
|             other ClientTarget#minifyCss                          6 ms
|          Target#rewriteSourceMaps                                 2 ms (2)
|          sha1                                                    52 ms (180)
|          other Target#make                                       53 ms
|    bundler.bundle..makeServerTarget...........................1,950 ms (1)
|       Target#make.............................................1,950 ms (1)
|          Isopack#getUnibuildAtArch                               10 ms (996)
|          Target#_runCompilerPlugins.............................428 ms (1)
|             files.realpath                                       73 ms (99)
|             sha1                                                  6 ms (284)
|             Isopack#getUnibuildAtArch                            12 ms (984)
|             plugin ecmascript...................................264 ms (2)
|                Babel.compile                                    182 ms (154)
|                sha1                                               3 ms (154)
|                other plugin ecmascript                           79 ms
|             plugin coffeescript...................................9 ms (1)
|                wrapped.fs.readFileSync                            1 ms (1)
|                other plugin coffeescript                          8 ms
|             other Target#_runCompilerPlugins                     63 ms
|          Target#_emitResources................................1,490 ms (1)
|             PackageSourceBatch.computeJsOutputFilesMap........1,083 ms (1)
|                ImportScanner#_getInstallPath                     18 ms (347)
|                ImportScanner#_findImportedModuleIdentifiers     479 ms (256)
|                ImportScanner#_tryToResolveImportedPath..........412 ms (549)
|                   files.stat                                    300 ms (1842)
|                   ImportScanner#_readFile........................17 ms (40)
|                      files.readFile                              13 ms (40)
|                      other ImportScanner#_readFile                3 ms
|                   other ImportScanner#_tryToResolveImportedPath  95 ms
|                ImportScanner#_readFile...........................81 ms (190)
|                   files.readFile                                 63 ms (190)
|                   sha1                                            4 ms (190)
|                   other ImportScanner#_readFile                  14 ms
|                other PackageSourceBatch.computeJsOutputFilesMap  93 ms
|             PackageSourceBatch#getResources.....................396 ms (100)
|                PackageSourceBatch#_linkJS.......................392 ms (100)
|                   sha1                                            3 ms (100)
|                   files.readFile                                202 ms (100)
|                   other PackageSourceBatch#_linkJS              187 ms
|                other PackageSourceBatch#getResources              4 ms
|             other Target#_emitResources                          11 ms
|          other Target#make                                       21 ms
|    bundler writeSiteArchive...................................1,361 ms (1)
|       files.rm_recursive                                          2 ms (1)
|       Builder#write...............................................5 ms (4)
|          files.writeFile                                          2 ms (4)
|          other Builder#write                                      1 ms
|       bundler writeTargetToPath...............................1,182 ms (3)
|          files.rm_recursive                                       2 ms (3)
|          files.stat                                               2 ms (11)
|          files.mkdir                                              1 ms (4)
|          ClientTarget#write.....................................669 ms (2)
|             files.mkdir                                           2 ms (10)
|             Builder#write.......................................205 ms (180)
|                Builder#_ensureDirectory                           2 ms (180)
|                sha1                                              23 ms (180)
|                files.writeFile                                   98 ms (180)
|                files.rename                                      48 ms (180)
|                other Builder#write                               33 ms
|             bundler writeFile...................................244 ms (208)
|                Builder#write....................................236 ms (208)
|                   Builder#_ensureDirectory                        5 ms (208)
|                   files.writeFile                               139 ms (208)
|                   files.rename                                   57 ms (208)
|                   other Builder#write                            35 ms
|                other bundler writeFile                            9 ms
|             files.writeFile                                       2 ms (2)
|             other ClientTarget#write                            215 ms
|          Builder#complete.........................................2 ms (3)
|             files.renameDirAlmostAtomically.......................2 ms (3)
|                files.rename                                       2 ms (6)
|          ServerTarget#write.....................................504 ms (1)
|             Builder#write........................................15 ms (7)
|                files.readFile                                     5 ms (12)
|                files.writeFile                                    4 ms (6)
|                files.rename                                       2 ms (6)
|                other Builder#write                                3 ms
|             JsImage#write.......................................452 ms (1)
|                files.mkdir                                        6 ms (27)
|                Builder#write....................................257 ms (203)
|                   Builder#_ensureDirectory                        3 ms (203)
|                   sha1                                           35 ms (203)
|                   files.writeFile                               117 ms (203)
|                   files.rename                                   62 ms (203)
|                   other Builder#write                            40 ms
|                Builder#copyDirectory.............................27 ms (21)
|                   files.symlink                                  24 ms (21)
|                   other Builder#copyDirectory                     3 ms
|                other JsImage#write                              161 ms
|             other ServerTarget#write                             36 ms
|          other bundler writeTargetToPath                          3 ms
|       Builder#complete..........................................166 ms (1)
|          files.renameDirAlmostAtomically........................166 ms (1)
|             files.rm_recursive                                  165 ms (1)
|       other bundler writeSiteArchive                              4 ms
|    other Build App                                               10 ms
| 
| Top leaves:
| ImportScanner#_findImportedModuleIdentifiers.............4,484 ms
| files.stat...............................................3,831 ms
| other plugin ecmascript..................................1,710 ms
| files.readFile...........................................1,303 ms
| other ImportScanner#_tryToResolveImportedPath............1,200 ms
| Babel.compile..............................................714 ms
| files.exists...............................................707 ms
| other runJavaScript packages/CosmosBrowserify_plugin.js....505 ms
| other PackageSourceBatch#_linkJS...........................451 ms
| other PackageSourceBatch.computeJsOutputFilesMap...........436 ms
| other runJavaScript packages/compileCoffeescript_plugin.js.375 ms
| files.writeFile............................................363 ms
| runJavaScript packages/underscore.js.......................333 ms
| files.realpath.............................................286 ms
| other compileUnibuild (the app)............................272 ms
| other runJavaScript packages/meteor.js.....................246 ms
| other Target#_runCompilerPlugins...........................234 ms
| other ClientTarget#write...................................215 ms
| sha1.......................................................199 ms
| runJavaScript packages/ecmascript-runtime.js...............192 ms
| files.rename...............................................173 ms
| files.rm_recursive.........................................170 ms
| other JsImage#write........................................161 ms
| other ImportScanner#_readFile..............................152 ms
| other runJavaScript packages/caching-compiler.js...........127 ms
| ImportScanner#_getInstallPath..............................126 ms
| runJavaScript packages/babel-runtime.js....................122 ms
| CssTools.stringifyCss......................................113 ms
| files.readdir..............................................113 ms
| other Builder#write........................................112 ms
| runJavaScript packages/html-tools.js.......................108 ms
| runJavaScript packages/ejson.js............................105 ms
| 
| (#2) Total: 21,366 ms (Build App)
| 
I20160205-15:33:52.487(-5)? Kadira: completed instrumenting the app
=> Started your app.                                                                
```

**Comment by markoshust** [comment]
And a couple reloads....


```

| (#3) Profiling: ProjectContext prepareProjectForBuild
|  Preparing project                         \
| ProjectContext prepareProjectForBuild...........................418 ms (1)
|    _initializeCatalog                                             4 ms (1)
|    _resolveConstraints...........................................59 ms (1)
|       Select Package Versions....................................46 ms (1)
|          new CS.Input                                             6 ms (1)
|          other Select Package Versions                           40 ms
|       other _resolveConstraints                                  12 ms
|    _downloadMissingPackages.....................................342 ms (1)
|       Isopack.readMetadataFromDirectory.........................189 ms (109)
|          files.exists                                            43 ms (109)
|          files.readFile                                         118 ms (109)
|          other Isopack.readMetadataFromDirectory                 27 ms
|       other _downloadMissingPackages                            154 ms
|    _buildLocalPackages...........................................12 ms (1)
|       IsopackCache Load local isopack.............................5 ms (109)
|          Isopack#getStrongOrderedUsedAndImpliedPackages           4 ms (109)
|          other IsopackCache Load local isopack                    1 ms
|       other _buildLocalPackages                                   6 ms
|    other ProjectContext prepareProjectForBuild                    2 ms
| 
| Top leaves:
| other _downloadMissingPackages.............................154 ms
| files.readFile.............................................118 ms
| 
| (#3) Total: 418 ms (ProjectContext prepareProjectForBuild)
| 
| (#4) Profiling: Rebuild App                \
|  Building the application                  /
| files.readFile                                                    8 ms (3)
| files.exists                                                      2 ms (4)
| files.writeFileAtomically.......................................434 ms (3)
|    files.writeFile                                              430 ms (3)
|    files.rename                                                   4 ms (3)
| Rebuild App..................................................13,399 ms (1)
|    initFromAppDir                                                 1 ms (1)
|    compiler.compile(the app)..................................1,699 ms (1)
|       compileUnibuild (the app)...............................1,697 ms (3)
|          Isopack#getUnibuildAtArch                                6 ms (189)
|          files.readdir                                          264 ms (555)
|          files.stat                                             725 ms (1793)
|          files.realpath                                         138 ms (278)
|          files.readFile                                         147 ms (363)
|          sha1                                                    10 ms (363)
|          other compileUnibuild (the app)                        406 ms
|       other compiler.compile(the app)                             2 ms
|    compiler.lint(the app).........................................4 ms (1)
|       Isopack#getUnibuildAtArch                                   2 ms (126)
|       other compiler.lint(the app)                                2 ms
|    Isopack#getUnibuildAtArch                                      2 ms (189)
|    bundler.bundle..makeClientTarget...........................9,851 ms (2)
|       Target#make.............................................9,851 ms (2)
|          Isopack#getUnibuildAtArch                               24 ms (1770)
|          Target#_runCompilerPlugins...........................3,203 ms (2)
|             files.stat                                           43 ms (2)
|             files.realpath                                      168 ms (174)
|             sha1                                                 14 ms (465)
|             Isopack#getUnibuildAtArch                            29 ms (1714)
|             plugin ecmascript.................................2,781 ms (4)
|                Babel.compile                                  2,574 ms (310)
|                sha1                                               7 ms (310)
|                other plugin ecmascript                          200 ms
|             plugin templating                                     3 ms (2)
|             plugin meteor                                         6 ms (2)
|             plugin cosmos:browserify                              3 ms (2)
|             plugin coffeescript                                   2 ms (2)
|             other Target#_runCompilerPlugins                    154 ms
|          Target#_emitResources................................6,335 ms (2)
|             PackageSourceBatch.computeJsOutputFilesMap........5,067 ms (2)
|                ImportScanner#_getInstallPath                     93 ms (1416)
|                ImportScanner#_findImportedModuleIdentifiers      53 ms (1344)
|                ImportScanner#_tryToResolveImportedPath........4,228 ms (4700)
|                   files.stat                                  3,156 ms (21654)
|                   ImportScanner#_readFile.......................263 ms (724)
|                      files.readFile                             213 ms (724)
|                      sha1                                        13 ms (712)
|                      other ImportScanner#_readFile               38 ms
|                   ImportScanner#_getInstallPath                   4 ms (42)
|                   other ImportScanner#_tryToResolveImportedPath 805 ms
|                ImportScanner#_readFile..........................438 ms (1098)
|                   files.readFile                                326 ms (1098)
|                   sha1                                           23 ms (1098)
|                   other ImportScanner#_readFile                  89 ms
|                other PackageSourceBatch.computeJsOutputFilesMap 255 ms
|             PackageSourceBatch#getResources...................1,261 ms (176)
|                PackageSourceBatch#_linkJS.....................1,258 ms (176)
|                   sha1                                            2 ms (176)
|                   linker.fullLink.............................1,176 ms (2)
|                      linker Module#getPrelinkedFiles..........1,172 ms (2)
|                         linker File#getPrelinkedOutput          407 ms (198)
|                         getPrelinkedFiles toStringWithSourceMap 751 ms (2)
|                         other linker Module#getPrelinkedFiles    14 ms
|                      other linker.fullLink                        3 ms
|                   other PackageSourceBatch#_linkJS               79 ms
|                other PackageSourceBatch#getResources              2 ms
|             other Target#_emitResources                           8 ms
|          Target#minifyJs                                         26 ms (2)
|          ClientTarget#minifyCss.................................168 ms (2)
|             mergeCss............................................162 ms (2)
|                CssTools.parseCss                                 64 ms (68)
|                CssTools.mergeCssAsts.............................12 ms (2)
|                   CssTools.rewriteCssUrls                         9 ms (68)
|                   other CssTools.mergeCssAsts                     3 ms
|                CssTools.stringifyCss                             45 ms (2)
|                composing source maps                             26 ms (2)
|                other mergeCss                                    15 ms
|             other ClientTarget#minifyCss                          7 ms
|          Target#rewriteSourceMaps                                 2 ms (2)
|          sha1                                                    51 ms (180)
|          other Target#make                                       42 ms
|    bundler.bundle..makeServerTarget...........................1,334 ms (1)
|       Target#make.............................................1,334 ms (1)
|          Isopack#getUnibuildAtArch                               13 ms (996)
|          Target#_runCompilerPlugins.............................520 ms (1)
|             files.stat                                           37 ms (1)
|             files.realpath                                       87 ms (99)
|             sha1                                                  6 ms (284)
|             Isopack#getUnibuildAtArch                            21 ms (984)
|             plugin ecmascript...................................291 ms (2)
|                Babel.compile                                    171 ms (154)
|                sha1                                               3 ms (154)
|                other plugin ecmascript                          117 ms
|             plugin coffeescript                                   2 ms (1)
|             other Target#_runCompilerPlugins                     75 ms
|          Target#_emitResources..................................772 ms (1)
|             PackageSourceBatch.computeJsOutputFilesMap..........647 ms (1)
|                ImportScanner#_getInstallPath                     21 ms (347)
|                ImportScanner#_findImportedModuleIdentifiers       2 ms (256)
|                ImportScanner#_tryToResolveImportedPath..........421 ms (549)
|                   files.stat                                    312 ms (1842)
|                   ImportScanner#_readFile........................22 ms (40)
|                      files.readFile                              18 ms (40)
|                      other ImportScanner#_readFile                2 ms
|                   other ImportScanner#_tryToResolveImportedPath  87 ms
|                ImportScanner#_readFile..........................105 ms (190)
|                   files.readFile                                 86 ms (190)
|                   sha1                                            4 ms (190)
|                   other ImportScanner#_readFile                  15 ms
|                other PackageSourceBatch.computeJsOutputFilesMap  97 ms
|             PackageSourceBatch#getResources.....................121 ms (100)
|                PackageSourceBatch#_linkJS.......................120 ms (100)
|                   sha1                                            1 ms (100)
|                   linker.fullLink...............................102 ms (1)
|                      linker Module#getPrelinkedFiles............101 ms (1)
|                         linker File#getPrelinkedOutput           38 ms (32)
|                         getPrelinkedFiles toStringWithSourceMap  60 ms (1)
|                         other linker Module#getPrelinkedFiles     3 ms
|                   other PackageSourceBatch#_linkJS               16 ms
|                other PackageSourceBatch#getResources              1 ms
|             other Target#_emitResources                           4 ms
|          other Target#make                                       28 ms
|    bundler writeSiteArchive.....................................503 ms (1)
|       files.exists                                                2 ms (1)
|       Builder#write...............................................2 ms (4)
|          files.readFile                                           2 ms (1)
|       bundler writeTargetToPath.................................496 ms (3)
|          ClientTarget#write.....................................248 ms (2)
|             Builder#write........................................49 ms (180)
|                Builder#_ensureDirectory                           2 ms (180)
|                sha1                                              27 ms (180)
|                files.writeFile                                    3 ms (2)
|                files.rename                                       4 ms (2)
|                other Builder#write                               13 ms
|             bundler writeFile....................................18 ms (208)
|                Builder#write.....................................16 ms (208)
|                   Builder#_ensureDirectory                        2 ms (208)
|                   files.writeFile                                 6 ms (2)
|                   files.rename                                    1 ms (2)
|                   other Builder#write                             6 ms
|                other bundler writeFile                            2 ms
|             files.writeFile                                       3 ms (2)
|             files.rename                                          2 ms (2)
|             other ClientTarget#write                            176 ms
|          Builder#complete                                         2 ms (3)
|          ServerTarget#write.....................................245 ms (1)
|             Builder#write.........................................6 ms (7)
|                files.readFile                                     3 ms (6)
|                files.rm_recursive                                 1 ms (1)
|             JsImage#write.......................................188 ms (1)
|                Builder#write.....................................36 ms (203)
|                   Builder#_ensureDirectory                        1 ms (203)
|                   sha1                                           27 ms (203)
|                   other Builder#write                             7 ms
|                Builder#copyDirectory.............................22 ms (21)
|                   files.symlink                                   9 ms (42)
|                   files.rm_recursive                             10 ms (21)
|                   other Builder#copyDirectory                     3 ms
|                other JsImage#write                              129 ms
|             other ServerTarget#write                             51 ms
|       other bundler writeSiteArchive                              2 ms
|    other Rebuild App                                              5 ms
| 
| Top leaves:
| files.stat...............................................4,274 ms
| Babel.compile............................................2,745 ms
| other ImportScanner#_tryToResolveImportedPath..............892 ms
| getPrelinkedFiles toStringWithSourceMap....................811 ms
| files.readFile.............................................804 ms
| linker File#getPrelinkedOutput.............................446 ms
| files.writeFile............................................444 ms
| other compileUnibuild (the app)............................406 ms
| files.realpath.............................................393 ms
| other PackageSourceBatch.computeJsOutputFilesMap...........352 ms
| other plugin ecmascript....................................317 ms
| files.readdir..............................................264 ms
| other Target#_runCompilerPlugins...........................229 ms
| sha1.......................................................192 ms
| other ClientTarget#write...................................176 ms
| other ImportScanner#_readFile..............................143 ms
| other JsImage#write........................................129 ms
| ImportScanner#_getInstallPath..............................118 ms
| 
| (#4) Total: 13,843 ms (Rebuild App)
| 
                                             |
                                             -
I20160205-15:35:26.567(-5)? Kadira: completed instrumenting the app
                                             |                          
                                              
| (#5) Profiling: ProjectContext prepareProjectForBuild
|  Checking for webapp-hashing@1.0.6-mod...  |
| ProjectContext prepareProjectForBuild...........................568 ms (1)
|    _initializeCatalog                                             4 ms (1)
|    _resolveConstraints..........................................187 ms (1)
|       Select Package Versions...................................175 ms (1)
|          new CS.Input                                            14 ms (1)
|          other Select Package Versions                          161 ms
|       other _resolveConstraints                                  11 ms
|    _downloadMissingPackages.....................................365 ms (1)
|       Isopack.readMetadataFromDirectory.........................195 ms (109)
|          files.exists                                            49 ms (109)
|          files.readFile                                         114 ms (109)
|          other Isopack.readMetadataFromDirectory                 32 ms
|       other _downloadMissingPackages                            170 ms
|    _buildLocalPackages...........................................11 ms (1)
|       IsopackCache Load local isopack.............................6 ms (109)
|          Isopack#getStrongOrderedUsedAndImpliedPackages           4 ms (109)
|          other IsopackCache Load local isopack                    2 ms
|       other _buildLocalPackages                                   5 ms
| 
| Top leaves:
| other _downloadMissingPackages.............................170 ms
| other Select Package Versions..............................161 ms
| files.readFile.............................................114 ms
| 
| (#5) Total: 568 ms (ProjectContext prepareProjectForBuild)
| 
| (#6) Profiling: Rebuild Appg@1.0.6-mod...  |
|  Building the application                  |
| files.readFile                                                    6 ms (3)
| files.exists                                                      2 ms (4)
| Rebuild App..................................................11,110 ms (1)
|    compiler.compile(the app)..................................1,443 ms (1)
|       compileUnibuild (the app)...............................1,442 ms (3)
|          Isopack#getUnibuildAtArch                                8 ms (189)
|          files.readdir                                          174 ms (555)
|          files.stat                                             617 ms (1793)
|          files.realpath                                          92 ms (278)
|          files.readFile                                         214 ms (363)
|          sha1                                                    11 ms (363)
|          other compileUnibuild (the app)                        326 ms
|       other compiler.compile(the app)                             1 ms
|    compiler.lint(the app).........................................5 ms (1)
|       Isopack#getUnibuildAtArch                                   2 ms (126)
|       other compiler.lint(the app)                                3 ms
|    Isopack#getUnibuildAtArch                                      2 ms (189)
|    bundler.bundle..makeClientTarget...........................7,981 ms (2)
|       Target#make.............................................7,981 ms (2)
|          Isopack#getUnibuildAtArch                               20 ms (1770)
|          Target#_runCompilerPlugins...........................1,037 ms (2)
|             files.realpath                                      188 ms (174)
|             Isopack#ensurePluginsInitialized                      2 ms (378)
|             sha1                                                 17 ms (465)
|             Isopack#getUnibuildAtArch                            30 ms (1714)
|             plugin ecmascript...................................624 ms (4)
|                Babel.compile                                    407 ms (310)
|                sha1                                               7 ms (310)
|                other plugin ecmascript                          210 ms
|             plugin templating                                     3 ms (2)
|             plugin meteor.........................................8 ms (2)
|                sha1                                               1 ms (64)
|                other plugin meteor                                7 ms
|             plugin cosmos:browserify                              3 ms (2)
|             plugin coffeescript                                   4 ms (2)
|             other Target#_runCompilerPlugins                    158 ms
|          Target#_emitResources................................6,429 ms (2)
|             PackageSourceBatch.computeJsOutputFilesMap........6,386 ms (2)
|                ImportScanner#_getInstallPath                    118 ms (1416)
|                ImportScanner#_findImportedModuleIdentifiers      15 ms (1344)
|                ImportScanner#_tryToResolveImportedPath........5,505 ms (4700)
|                   files.stat                                  4,079 ms (21654)
|                   ImportScanner#_readFile.......................356 ms (724)
|                      files.readFile                             280 ms (724)
|                      sha1                                        16 ms (712)
|                      other ImportScanner#_readFile               60 ms
|                   ImportScanner#_getInstallPath                   3 ms (42)
|                   other ImportScanner#_tryToResolveImportedPath 1,067 ms
|                ImportScanner#_readFile..........................459 ms (1098)
|                   files.readFile                                370 ms (1098)
|                   sha1                                           25 ms (1098)
|                   other ImportScanner#_readFile                  65 ms
|                other PackageSourceBatch.computeJsOutputFilesMap 288 ms
|             PackageSourceBatch#getResources......................32 ms (176)
|                PackageSourceBatch#_linkJS........................29 ms (176)
|                   sha1                                            4 ms (176)
|                   other PackageSourceBatch#_linkJS               25 ms
|                other PackageSourceBatch#getResources              3 ms
|             other Target#_emitResources                          11 ms
|          Target#minifyJs                                        127 ms (2)
|          ClientTarget#minifyCss.................................258 ms (2)
|             mergeCss............................................252 ms (2)
|                CssTools.parseCss                                 90 ms (68)
|                CssTools.mergeCssAsts.............................15 ms (2)
|                   CssTools.rewriteCssUrls                        12 ms (68)
|                   other CssTools.mergeCssAsts                     3 ms
|                CssTools.stringifyCss                             88 ms (2)
|                composing source maps                             33 ms (2)
|                other mergeCss                                    26 ms
|             other ClientTarget#minifyCss                          6 ms
|          Target#rewriteSourceMaps                                 1 ms (2)
|          sha1                                                    75 ms (180)
|          other Target#make                                       34 ms
|    bundler.bundle..makeServerTarget...........................1,247 ms (1)
|       Target#make.............................................1,247 ms (1)
|          Isopack#getUnibuildAtArch                               20 ms (996)
|          Target#_runCompilerPlugins.............................552 ms (1)
|             files.realpath                                      111 ms (99)
|             sha1                                                  7 ms (284)
|             Isopack#getUnibuildAtArch                            17 ms (984)
|             plugin ecmascript...................................349 ms (2)
|                Babel.compile                                    227 ms (154)
|                sha1                                               5 ms (154)
|                other plugin ecmascript                          118 ms
|             plugin coffeescript                                   1 ms (1)
|             other Target#_runCompilerPlugins                     66 ms
|          Target#_emitResources..................................656 ms (1)
|             PackageSourceBatch.computeJsOutputFilesMap..........642 ms (1)
|                ImportScanner#_getInstallPath                     18 ms (347)
|                ImportScanner#_findImportedModuleIdentifiers       2 ms (256)
|                ImportScanner#_tryToResolveImportedPath..........395 ms (549)
|                   files.stat                                    289 ms (1842)
|                   ImportScanner#_readFile........................35 ms (40)
|                      files.readFile                              32 ms (40)
|                      other ImportScanner#_readFile                2 ms
|                   other ImportScanner#_tryToResolveImportedPath  70 ms
|                ImportScanner#_readFile..........................161 ms (190)
|                   files.readFile                                138 ms (190)
|                   sha1                                            5 ms (190)
|                   other ImportScanner#_readFile                  18 ms
|                other PackageSourceBatch.computeJsOutputFilesMap  66 ms
|             PackageSourceBatch#getResources.......................9 ms (100)
|                PackageSourceBatch#_linkJS.........................8 ms (100)
|                   sha1                                            1 ms (100)
|                   other PackageSourceBatch#_linkJS                6 ms
|                other PackageSourceBatch#getResources              2 ms
|             other Target#_emitResources                           4 ms
|          other Target#make                                       19 ms
|    bundler writeSiteArchive.....................................419 ms (1)
|       Builder#write                                               1 ms (4)
|       bundler writeTargetToPath.................................415 ms (3)
|          ClientTarget#write.....................................195 ms (2)
|             Builder#write........................................31 ms (180)
|                Builder#_ensureDirectory                           1 ms (180)
|                sha1                                              19 ms (180)
|                files.writeFile                                    2 ms (2)
|                other Builder#write                                8 ms
|             bundler writeFile....................................14 ms (208)
|                Builder#write.....................................11 ms (208)
|                   Builder#_ensureDirectory                        2 ms (208)
|                   files.writeFile                                 3 ms (2)
|                   other Builder#write                             6 ms
|                other bundler writeFile                            2 ms
|             files.writeFile                                       1 ms (2)
|             other ClientTarget#write                            149 ms
|          Builder#complete                                         2 ms (3)
|          ServerTarget#write.....................................218 ms (1)
|             Builder#write.........................................7 ms (7)
|                files.readFile                                     5 ms (6)
|             JsImage#write.......................................170 ms (1)
|                Builder#write.....................................32 ms (203)
|                   Builder#_ensureDirectory                        2 ms (203)
|                   sha1                                           23 ms (203)
|                   other Builder#write                             8 ms
|                Builder#copyDirectory.............................22 ms (21)
|                   files.symlink                                   8 ms (42)
|                   files.rm_recursive                             10 ms (21)
|                   other Builder#copyDirectory                     3 ms
|                other JsImage#write                              116 ms
|             other ServerTarget#write                             39 ms
|       other bundler writeSiteArchive                              2 ms
|    other Rebuild App                                             12 ms
| 
| Top leaves:
| files.stat...............................................4,986 ms
| other ImportScanner#_tryToResolveImportedPath............1,137 ms
| files.readFile...........................................1,046 ms
| Babel.compile..............................................633 ms
| files.realpath.............................................391 ms
| other PackageSourceBatch.computeJsOutputFilesMap...........354 ms
| other plugin ecmascript....................................328 ms
| other compileUnibuild (the app)............................326 ms
| other Target#_runCompilerPlugins...........................223 ms
| sha1.......................................................219 ms
| files.readdir..............................................174 ms
| other ClientTarget#write...................................149 ms
| other ImportScanner#_readFile..............................145 ms
| ImportScanner#_getInstallPath..............................139 ms
| Target#minifyJs............................................127 ms
| other JsImage#write........................................116 ms
| 
| (#6) Total: 11,118 ms (Rebuild App)
| 
I20160205-15:35:44.129(-5)? Kadira: completed instrumenting the app
```

**Comment by dgreensp** [comment]
Thanks!  Maybe you should paste it into https://gist.github.com/ next time for aesthetic purposes.

Ok, that matches what I've been seeing.  beta.8 (coming soon) should shave a few seconds off of rebuild in `ImportScanner` and a few seconds off start-up in `Select Package Versions`!

**Comment by markoshust** [comment]
ok great will check that out. sorry about the posts, i'll put longer snippets in a gist nextime.

**Comment by dgreensp** [comment]
see also: https://github.com/meteor/meteor/blob/release-1.3/tools/PERFORMANCE.md

**Comment by markoshust** [comment]
getting better times on 1.3b8. reloads now getting around 10 sec or less... this is much better :)

https://gist.github.com/markoshust/586a5c7587474d96d571

**Comment by clayne11** [comment]
I'm running 1.2 and I just profile my app for the first time.

On rebuilds it's taking between 7-9 seconds for "other ClientTarget#minifyCss" even if I don't make a CSS change!

CSS files should not be getting rebuilt unless they are touched. This is killing my build time really badly.

**Comment by veered** [comment]
Check out https://github.com/meteor/meteor/pull/5875. It would be really cool if we could get something like this merged. We will probably have to go the caching route.

**Comment by c9s** [comment]
I waded into the isopacker source code, I would suggest these large functions should be break into small monomorphic functions to let V8 optimize.

there are few requirements to let v8 optimize the function:

- with fixed argument length.
- with fixed argument types.
- functions smaller than ~50 lines.

**Comment by hwillson** [comment]
Meteor [1.4.2](https://github.com/meteor/meteor/blob/devel/History.md#v142) has introduced many build tool performance improvements, some of which were the direct result of this issue thread. As part of the on-going old issue cleanup process, the time has come to close this thread. If anyone here has further build tool performance improvement suggestions, please log those suggestions in new issues. A **BIG** thanks to everyone involved!
